[git-webkit] Timing out IPV6 requests

JonWBedard opened the pull request:
#### c06388334ce2311ed2c6ac0cdde483f9ed6971e8
<pre>
[git-webkit] Timing out IPV6 requests
<a href="https://bugs.webkit.org/show_bug.cgi?id=263124">https://bugs.webkit.org/show_bug.cgi?id=263124</a>
rdar://116923729

Reviewed by Dewei Zhu.

Because Python has not implemented the &quot;Happy Eyeballs&quot; algorithm racing
IPV4 and IPV6 connections, we want to disable IPV6 once we know that IPV6
stack on this machine is broken. We do this by overriding urllib3&apos;s
create_connection function.

* Tools/Scripts/webkitpy/__init__.py:
(create_connection): Disable IPV6 if our first request to an IPV6 address fails.

Canonical link: <a href="https://commits.webkit.org/269440@main">https://commits.webkit.org/269440@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/91afa51c3e05fa6c76157069b1d97206719f712f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22583 "3 style errors") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23667 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24489 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/20906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22843 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/1316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/23114 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22822 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/1316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/19599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25342 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/22767 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/1316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/20463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/1316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/20707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/24544 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/23114 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2839 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

dewei-zhu reviewed (approved):
r=me

dewei-zhu commented on Tools/Scripts/webkitpy/__init__.py:
```diff
@@ -118,3 +133,77 @@
+        if af == socket.AF_INET6:
+            if not urllib3.util.connection.HAS_IPV6:
+                continue
+            HAS_CHECKED_IPV6 = True
```
Nit: I feel we set this to True prematurely based on the variable name. Also, is this thread-safe?

dewei-zhu reviewed (approved):
r=me

JonWBedard committed:
[git-webkit] Timing out IPV6 requests
https://bugs.webkit.org/show_bug.cgi?id=263124
rdar://116923729

Reviewed by Dewei Zhu.

Because Python has not implemented the "Happy Eyeballs" algorithm racing
IPV4 and IPV6 connections, we want to disable IPV6 once we know that IPV6
stack on this machine is broken. We do this by overriding urllib3's
create_connection function.

* Tools/Scripts/webkitpy/__init__.py:
(create_connection): Disable IPV6 if our first request to an IPV6 address fails.

Canonical link: https://commits.webkit.org/269440@main
